### PR TITLE
Draft: Bullseye->Bookworm migration

### PR DESCRIPTION
**Pull request by alexAubin**
cf https://github.com/YunoHost/issues/issues/2050

### Notes to test migration on a fresh VPS

```bash
curl https://install.yunohost.org | bash -s -- -a
# Drink coffee
yunohost tools postinstall -d bullseye-bookworm-migration.test -u camille -F Camille -p Yunohost
# Sip more coffee

# Install your favorite apps, for example:
# Nextcloud (PHP+mysql)
yunohost app install nextcloud --args "domain=bullseye-bookworm-migration.test&path=/cloud&admin=camille&init_main_permission=visitors&user_home=true&system_addressbook_exposed=yes"
# Wekan (Nodejs + mongodb) 
yunohost app install wekan --args "domain=bullseye-bookworm-migration.test&path=/wekan&admin=camille&init_main_permission=visitors"
# Terraforming mars (nodejs+postgresql) Actually requires a few GB of RAM to npm build ? :|
# yunohost app install terraforming-mars --args "domain=bullseye-bookworm-migration.test&path=/tfm&init_main_permission=visitors"
# Coin (Python + postgresql)
yunohost domain add coin.bullseye-bookworm-migration.test
yunohost app install coin --args "domain=coin.bullseye-bookworm-migration.test&admin=camille&email=camille@example.tld&isp_name=YunoHost&isp_site=yunohost.org"

# Gotta make sure we get to unstable, because only unstable packages of yunohost 12.x are available now
sed -i 's/stable/stable testing unstable/g' /etc/apt/sources.list.d/yunohost.list

# Dirty/manual apply of this PR
apt install aptitude -y
curl https://raw.githubusercontent.com/YunoHost/yunohost/migrate-to-bookworm/locales/en.json > /usr/share/yunohost/locales/en.json
curl https://raw.githubusercontent.com/YunoHost/yunohost/migrate-to-bookworm/src/utils/system.py > /usr/lib/python3/dist-packages/yunohost/utils/system.py
curl https://raw.githubusercontent.com/YunoHost/yunohost/migrate-to-bookworm/src/migrations/0027_migrate_to_bookworm.py > /usr/lib/python3/dist-packages/yunohost/migrations/0027_migrate_to_bookworm.py

# Disable check that system is up to date before migration
sed -i 's/raise.*not_fully_up_to_date.*/pass/g' /usr/lib/python3/dist-packages/yunohost/migrations/0027_migrate_to_bookworm.py

# Validate the migration is there
yunohost tools migrations list --pending

# Run the migration from the command line (OR restart yunohost-api and run the migration from the webadmin?)
yunohost tools migrations run --accept-disclaimer --debug

# Here be dragons
```

### CODE REVIEW

**Comment by alexAubin** [comment]
So it's working~ish on a clean install, next step is to check upgrading with a bunch of apps installed, trying to trigger tricky cases, polishing

File: src/migrations/0027_migrate_to_bookworm.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+        os.system(
+            "LC_ALL=C DEBIAN_FRONTEND=noninteractive APT_LISTCHANGES_FRONTEND=none apt autoremove --assume-yes"
+        )
+        os.system("apt clean --assume-yes")
+
+        #
+        # Stupid hack for stupid dnsmasq not picking up its new init.d script then breaking everything ...
+        # https://forum.yunohost.org/t/20676
+        #
+        # FIXME : this is from buster->bullseye, do we still needed it ?
```
Comment (kay0u) [inline src/migrations/0027_migrate_to_bookworm.py:197]: To be checked during the testing phase

File: src/migrations/0027_migrate_to_bookworm.py [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+        #if os.system(cmd) == 0:
+        #    raise YunohostError(
+        #        "The upgrade cannot be completed, because some app dependencies would need to be removed?",
+        #        raw_msg=True,
+        #    )
+
+        # FIXME :
+        #postupgradecmds = "rm -f /usr/sbin/policy-rc.d\n"
+        #postupgradecmds += "echo 'Restarting nginx...' >&2\n"
+        #postupgradecmds += "systemctl restart nginx\n"
```
Comment (kay0u) [inline src/migrations/0027_migrate_to_bookworm.py:231]: Wut, can we remove that?

File: src/migrations/0027_migrate_to_bookworm.py [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+        #
+
+        # We only patch the conflict if we're on yunohost 11.x
+        if self.yunohost_major_version() != N_CURRENT_YUNOHOST:
+            return
+
+        conflicts = check_output("dpkg-query -s yunohost | grep '^Conflicts:'").strip()
+        if conflicts:
+            # We want to keep conflicting with apache/bind9 tho
+            new_conflicts = "Conflicts: apache2, bind9"
```
Comment (kay0u) [inline src/migrations/0027_migrate_to_bookworm.py:421]: What about other conflicts ? In yunohost, admin, ssowat etc?

**Comment by tituspijean** [comment]
I have just upgraded a server from YunoHost 11.2.11.3 to 12 after following the instructions of the first message.

Migration runs smoothly, but the logging has a bug:

```
...
690523 SUCCESS Migration 0027_migrate_to_bookworm completed
690547 DEBUG action [68666.1] executed in 690.171s
690548 DEBUG lock has been released
Traceback (most recent call last):
  File "/usr/bin/yunohost", line 77, in <module>
    yunohost.cli(
  File "/usr/lib/python3/dist-packages/yunohost/__init__.py", line 41, in cli
    ret = moulinette.cli(
  File "/usr/lib/python3/dist-packages/moulinette/__init__.py", line 110, in cli
    load_only_category = args[0] if args and not args[0].startswith("-") else None
  File "/usr/lib/python3/dist-packages/moulinette/interfaces/cli.py", line 503, in run
    raise MoulinetteError("operation_interrupted")
  File "/usr/lib/python3/dist-packages/moulinette/actionsmap.py", line 580, in process
    - top_parser -- A BaseActionsMapParser-derived instance to use for
  File "/usr/lib/python3/dist-packages/yunohost/tools.py", line 825, in tools_migrations_run
    return sorted(migrations, key=lambda m: m.id)
  File "/usr/lib/python3/dist-packages/yunohost/log.py", line 671, in success
    """
  File "/usr/lib/python3/dist-packages/yunohost/log.py", line 709, in close
    msg = m18n.n("log_link_to_log", name=self.name, desc=desc)
  File "/usr/lib/python3/dist-packages/yunohost/log.py", line 797, in _get_description_from_name
    args = parts[1:]
ModuleNotFoundError: No module named '_strptime'
```

**Comment by centralscrutinizer21** [comment]
ok since I broke my matrix server for the n time I am going to be bad with my server and try this upgrade with several apps installed:
- calibreweb
- bludit
- bookwyrm
- castopod
- cinny
- collabora
- coturn
- element-call
- invidious
- libreddit
- mautrix_signal
- mautrix_telegram
- mautrix_whatsapp
- miniflux
- nextcloud
- yourls
- pleroma
- peertube
- my_webapp
- pixelfed
- rss-bridge
- navidrome
- umami
- vaultwarden
- wallabag2
- webtrees
- wireguard
- searxng
- ntfy

It's a sacrifice for the greater good. 
I have backups for the most important apps and media files in separate storages
I'll report the results

**Comment by Salamandar** [comment]
> I have just upgraded a server from YunoHost 11.2.11.3 to 12 after following the instructions of the first message.
> 
> Migration runs smoothly, but the logging has a bug:
> 

Same, and nginx is left dead

```
Jun 26 17:45:22 intra.syslinbit.com systemd[1]: Reload failed for nginx.service - A high performance web server and a reverse proxy server.
```

Restarting the nginx service works.

**Comment by Salamandar** [comment]
Webadmin issue : Clicking on this button changes the URL visible in the browser but doesn't actually redirect to the page…
Capture d’écran du 2024-07-04 19-41-32

File: src/migrations/0027_migrate_to_bookworm.py [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+        #        raw_msg=True,
+        #    )
+
+        # FIXME :
+        #postupgradecmds = "rm -f /usr/sbin/policy-rc.d\n"
+        #postupgradecmds += "echo 'Restarting nginx...' >&2\n"
+        #postupgradecmds += "systemctl restart nginx\n"
+
+        # If running from the webadmin, restart the API after a delay
+        if Moulinette.interface.type == "api":
```
Comment (Salamandar) [inline src/migrations/0027_migrate_to_bookworm.py:240]: Should be `moulinette` not `Moulinette`

File: src/utils/system.py [hunk 5]
```diff
@@ -211,3 +211,118 @@ def _dump_sources_list():
...
+        lambda l: log_apt_status_to_progress_bar(l.rstrip()),
+    )
+
+    cmd = (
+        f'LC_ALL=C DEBIAN_FRONTEND=noninteractive APT_LISTCHANGES_FRONTEND=none aptitude {cmd} --quiet=2 -o=Dpkg::Use-Pty=0 -o "APT::Status-Fd=$YNH_STDINFO"'
+    )
+
+    # If upgrading yunohost from the API, delay the Yunohost-api restart
+    # (this should be the last time we need it before bookworm, because on bookworm, yunohost-admin cookies will be persistent upon api restart)
+    if " yunohost " in cmd and Moulinette.interface.type == "api":
```
Comment (Salamandar) [inline src/utils/system.py:292]: Should be `moulinette` not `Moulinette`

**Comment by Salamandar** [comment]
As is, this PR can be merged without risk.

**Comment by alexAubin** [comment]
https://www.youtube.com/watch?v=AWM5ZNdWlqw

**Pull request merged**